// programmr_javamenu/lib/browser/programmr_javamenu-contribution.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
var __param = (this && this.__param) || function (paramIndex, decorator) {
    return function (target, key) { decorator(target, key, paramIndex); }
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
var __read = (this && this.__read) || function (o, n) {
    var m = typeof Symbol === "function" && o[Symbol.iterator];
    if (!m) return o;
    var i = m.call(o), r, ar = [], e;
    try {
        while ((n === void 0 || n-- > 0) && !(r = i.next()).done) ar.push(r.value);
    }
    catch (error) { e = { error: error }; }
    finally {
        try {
            if (r && !r.done && (m = i["return"])) m.call(i);
        }
        finally { if (e) throw e.error; }
    }
    return ar;
};
var __spread = (this && this.__spread) || function () {
    for (var ar = [], i = 0; i < arguments.length; i++) ar = ar.concat(__read(arguments[i]));
    return ar;
};
Object.defineProperty(exports, "__esModule", { value: true });
var inversify_1 = require("inversify");
var common_1 = require("@theia/core/lib/common");
var browser_1 = require("@theia/core/lib/browser");
var core_1 = require("@theia/core");
var file_resource_1 = require("@theia/filesystem/lib/browser/file-resource");
var browser_2 = require("@theia/workspace/lib/browser");
var filesystem_watcher_1 = require("@theia/filesystem/lib/browser/filesystem-watcher");
var common_2 = require("@theia/filesystem/lib/common");
var uri_1 = require("@theia/core/lib/common/uri");
var terminal_service_1 = require("@theia/terminal/lib/browser/base/terminal-service");
var frontend_application_state_1 = require("@theia/core/lib/browser/frontend-application-state");
exports.Programmr_javamenuCommand = {
    id: 'Programmr_javamenu.command',
    label: "Shows a message"
};
exports.RunExtensionCommand = {
    id: 'RunExtension.command',
    label: "Run"
};
exports.CompileExtensionCommand = {
    id: 'CompileExtension.command',
    label: "Compile"
};
var RunMenus;
(function (RunMenus) {
    RunMenus.RUN = __spread(core_1.MAIN_MENU_BAR, ['7_run']);
    RunMenus.RUN_NEW = __spread(RunMenus.RUN, ['3_run']);
    RunMenus.BUILD_NEW = __spread(RunMenus.RUN, ['1_run']);
    RunMenus.RUN_TASKS = __spread(RunMenus.RUN, ['2_run']);
    RunMenus.OPEN_NEW = __spread(RunMenus.RUN, ['4_run']);
})(RunMenus = exports.RunMenus || (exports.RunMenus = {}));
var Programmr_javamenuCommandContribution = /** @class */ (function () {
    function Programmr_javamenuCommandContribution(messageService, terminalService, fileSystem, watcher) {
        this.messageService = messageService;
        this.terminalService = terminalService;
        this.fileSystem = fileSystem;
        this.watcher = watcher;
        this.configObj = '';
    }
    Programmr_javamenuCommandContribution.prototype.setConfigObject = function (obj) {
        this.configObj = obj;
    };
    Programmr_javamenuCommandContribution.prototype.getConfigObject = function () {
        return this.configObj;
    };
    Programmr_javamenuCommandContribution.prototype.openDefaultFile = function (command) {
        return __awaiter(this, void 0, void 0, function () {
            var current, uri, resource, contents, e_1;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        _a.trys.push([0, 2, , 3]);
                        current = this.workspaceService.workspace;
                        uri = new uri_1.default(current.uri + "/.pmainfile");
                        resource = new file_resource_1.FileResource(uri, this.fileSystem, this.watcher);
                        return [4 /*yield*/, resource.readContents()];
                    case 1:
                        contents = _a.sent();
                        console.log("|" + contents + "|");
                        contents = contents.replace(/\n/g, '');
                        contents = contents.replace(/\r/g, '');
                        this.configObj = JSON.parse(contents);
                        if (command == 'run') {
                            if ("runmenucommand" in this.configObj) {
                                console.log("cmd entry found");
                                this.runTerminalCommand(this.configObj.runmenucommand);
                            }
                        }
                        else if (command == 'compile') {
                            if ("compilemenucommand" in this.configObj) {
                                console.log("cmd entry found");
                                this.runTerminalCommand(this.configObj.compilemenucommand);
                            }
                        }
                        return [3 /*break*/, 3];
                    case 2:
                        e_1 = _a.sent();
                        console.log(e_1);
                        return [3 /*break*/, 3];
                    case 3: return [2 /*return*/];
                }
            });
        });
    };
    Programmr_javamenuCommandContribution.prototype.registerCommands = function (registry) {
        var _this = this;
        console.log("after");
        registry.registerCommand(exports.CompileExtensionCommand, {
            execute: function () {
                _this.messageService.info('Compiling...');
                _this.openDefaultFile('compile');
            }
        });
        registry.registerCommand(exports.RunExtensionCommand, {
            execute: function () {
                _this.messageService.info('Running...');
                _this.openDefaultFile('run');
            }
        });
    };
    Programmr_javamenuCommandContribution.prototype.runTerminalCommand = function (command) {
        return __awaiter(this, void 0, void 0, function () {
            var terminal;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        terminal = this.terminalService.currentTerminal;
                        if (!!terminal) return [3 /*break*/, 3];
                        return [4 /*yield*/, this.terminalService.newTerminal({ created: new Date().toString() })];
                    case 1:
                        terminal = (_a.sent());
                        return [4 /*yield*/, terminal.start()];
                    case 2:
                        _a.sent();
                        this.terminalService.activateTerminal(terminal);
                        _a.label = 3;
                    case 3:
                        terminal.sendText(command + "\n");
                        return [2 /*return*/];
                }
            });
        });
    };
    __decorate([
        inversify_1.inject(frontend_application_state_1.FrontendApplicationStateService),
        __metadata("design:type", frontend_application_state_1.FrontendApplicationStateService)
    ], Programmr_javamenuCommandContribution.prototype, "stateService", void 0);
    __decorate([
        inversify_1.inject(browser_2.WorkspaceService),
        __metadata("design:type", browser_2.WorkspaceService)
    ], Programmr_javamenuCommandContribution.prototype, "workspaceService", void 0);
    Programmr_javamenuCommandContribution = __decorate([
        inversify_1.injectable(),
        __param(0, inversify_1.inject(common_1.MessageService)),
        __param(1, inversify_1.inject(terminal_service_1.TerminalService)),
        __param(2, inversify_1.inject(common_2.FileSystem)),
        __param(3, inversify_1.inject(filesystem_watcher_1.FileSystemWatcher)),
        __metadata("design:paramtypes", [common_1.MessageService, Object, Object, filesystem_watcher_1.FileSystemWatcher])
    ], Programmr_javamenuCommandContribution);
    return Programmr_javamenuCommandContribution;
}());
exports.Programmr_javamenuCommandContribution = Programmr_javamenuCommandContribution;
var Programmr_javamenuMenuContribution = /** @class */ (function () {
    function Programmr_javamenuMenuContribution() {
    }
    Programmr_javamenuMenuContribution.prototype.registerMenus = function (menus) {
        menus.registerSubmenu(RunMenus.RUN, 'Run');
        menus.registerMenuAction(RunMenus.BUILD_NEW, {
            commandId: exports.CompileExtensionCommand.id,
            label: 'Compile',
            order: '0'
        });
        menus.registerMenuAction(RunMenus.RUN_NEW, {
            commandId: exports.RunExtensionCommand.id,
            label: 'Run',
            order: '1'
        });
        menus.registerMenuAction(browser_1.CommonMenus.EDIT_FIND, {
            commandId: exports.RunExtensionCommand.id,
            label: 'Run'
        });
        menus.registerMenuAction(browser_1.CommonMenus.EDIT_FIND, {
            commandId: exports.CompileExtensionCommand.id,
            label: 'Compile'
        });
    };
    Programmr_javamenuMenuContribution = __decorate([
        inversify_1.injectable()
    ], Programmr_javamenuMenuContribution);
    return Programmr_javamenuMenuContribution;
}());
exports.Programmr_javamenuMenuContribution = Programmr_javamenuMenuContribution;
