// @theia/scm/lib/browser/scm-group-command-registry.js
"use strict";
/********************************************************************************
 * Copyright (C) 2019 Red Hat, Inc. and others.
 *
 * This program and the accompanying materials are made available under the
 * terms of the Eclipse Public License v. 2.0 which is available at
 * http://www.eclipse.org/legal/epl-2.0.
 *
 * This Source Code may also be made available under the following Secondary
 * Licenses when the conditions for such availability set forth in the Eclipse
 * Public License v. 2.0 are satisfied: GNU General Public License, version 2
 * with the GNU Classpath Exception which is available at
 * https://www.gnu.org/software/classpath/license.html.
 *
 * SPDX-License-Identifier: EPL-2.0 OR GPL-2.0 WITH Classpath-exception-2.0
 ********************************************************************************/
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
var __values = (this && this.__values) || function (o) {
    var m = typeof Symbol === "function" && o[Symbol.iterator], i = 0;
    if (m) return m.call(o);
    return {
        next: function () {
            if (o && i >= o.length) o = void 0;
            return { value: o && o[i++], done: !o };
        }
    };
};
Object.defineProperty(exports, "__esModule", { value: true });
var inversify_1 = require("inversify");
var core_1 = require("@theia/core");
exports.ScmGroupCommandContribution = Symbol('ScmGroupCommandContribution');
var ScmGroupCommandRegistry = /** @class */ (function () {
    function ScmGroupCommandRegistry() {
        this.items = new Map();
    }
    ScmGroupCommandRegistry.prototype.onStart = function () {
        var e_1, _a;
        var contributions = this.contributionProvider.getContributions();
        try {
            for (var contributions_1 = __values(contributions), contributions_1_1 = contributions_1.next(); !contributions_1_1.done; contributions_1_1 = contributions_1.next()) {
                var contribution = contributions_1_1.value;
                contribution.registerScmGroupCommands(this);
            }
        }
        catch (e_1_1) { e_1 = { error: e_1_1 }; }
        finally {
            try {
                if (contributions_1_1 && !contributions_1_1.done && (_a = contributions_1.return)) _a.call(contributions_1);
            }
            finally { if (e_1) throw e_1.error; }
        }
    };
    ScmGroupCommandRegistry.prototype.registerItems = function (groupId, items) {
        var savedItems = this.items.get(groupId);
        if (savedItems) {
            items.forEach(function (item) { return savedItems.push(item); });
            this.items.set(groupId, savedItems);
        }
        else {
            this.items.set(groupId, items);
        }
    };
    ScmGroupCommandRegistry.prototype.registerItem = function (groupId, item) {
        var items = this.items.get(groupId);
        if (items) {
            items.push(item);
            this.items.set(groupId, items);
        }
        else {
            this.items.set(groupId, [item]);
        }
    };
    ScmGroupCommandRegistry.prototype.getItems = function (groupId) {
        return this.items.get(groupId);
    };
    __decorate([
        inversify_1.inject(core_1.ContributionProvider),
        inversify_1.named(exports.ScmGroupCommandContribution),
        __metadata("design:type", Object)
    ], ScmGroupCommandRegistry.prototype, "contributionProvider", void 0);
    ScmGroupCommandRegistry = __decorate([
        inversify_1.injectable()
    ], ScmGroupCommandRegistry);
    return ScmGroupCommandRegistry;
}());
exports.ScmGroupCommandRegistry = ScmGroupCommandRegistry;
